pycontrails/physics/geo.py

```diff
@@ -853,3 +853,125 @@ def _dt_to_float_seconds(dt: np.ndarray | np.timedelta64, dtype: npt.DTypeLike)
+    latitude : np.ndarray
+        1D Latitude values with index corresponding to latitude inputs, [:math:`\deg`]
+    buffer: float
+        Add buffer to rectangular spatial bounding box, [:math:`\deg`]
+
+    Returns
+    -------
+    list[float]
+        Spatial bounding box, [lon_min, lat_min, lon_max, lat_max], [:math:`\deg`]
+    """
+    lon_min = max(np.floor(np.min(longitude) - buffer), -180)
+    lon_max = min(np.ceil(np.max(longitude) + buffer), 179.99)
+    lat_min = max(np.floor(np.min(latitude) - buffer), -90)
+    lat_max = min(np.ceil(np.max(latitude) + buffer), 90)
+    return [lon_min, lat_min, lon_max, lat_max]
+
+
+def domain_surface_area(
+        spatial_bbox: list[float] = [-180, -90, 180, 90], *,
+        spatial_grid_res: float = 0.5
+) -> float:
+    """
+    Calculate surface area in the provided spatial bounding box.
+
+    Parameters
+    ----------
+    spatial_bbox : list[float]
+        Spatial bounding box, [lon_min, lat_min, lon_max, lat_max], [:math:`\deg`]
+    spatial_grid_res : float
+        Spatial grid resolution, [:math:`\deg`]
+
+    Returns
+    -------
+    float
+        Domain surface area, [:math:`m^{2}`]
+    """
+    lon_coords = np.arange(spatial_bbox[0], spatial_bbox[2] + 0.01, spatial_grid_res)
+    lat_coords = np.arange(spatial_bbox[1], spatial_bbox[3] + 0.01, spatial_grid_res)
+    da_surface_area = grid_surface_area(lon_coords, lat_coords)
+    return np.nansum(da_surface_area)
+
+
+def grid_surface_area(
+        lon_coords: npt.NDArray[np.float_],
+        lat_coords: npt.NDArray[np.float_]
+) -> xr.DataArray:
+    """
+    Calculate surface area that is covered by each pixel in a longitude-latitude grid.
+
+    Parameters
+    ----------
+    lon_coords: npt.NDArray[np.float_]
+        Longitude coordinates in a longitude-latitude grid, [:math:`\deg`]
+    lat_coords: npt.NDArray[np.float_]
+        Latitude coordinates in a longitude-latitude grid, [:math:`\deg`]
+
+    Returns
+    -------
+    xr.DataArray
+        Surface area of each pixel in a longitude-latitude grid, [:math:`m^{2}`]
+
+    References
+    ----------
+    - https://www.pmel.noaa.gov/maillists/tmap/ferret_users/fu_2004/msg00023.html
+    """
+    # Ensure that grid spacing is uniform
+    d_lon = np.diff(lon_coords)
+    if np.all(d_lon == d_lon[0]) is False:
+        raise AssertionError("Longitude grid spacing is not uniform. ")
+
+    d_lat = np.diff(lat_coords)
+    if np.all(d_lat == d_lat[0]) is False:
+        raise AssertionError("Latitude grid spacing is not uniform. ")
+
+    lon_2d, lat_2d = np.meshgrid(lon_coords, lat_coords)
+    area_lat_btm = _area_between_latitude_and_north_pole(lat_2d - np.abs(d_lat[0]))
+    area_lat_top = _area_between_latitude_and_north_pole(lat_2d)
+    area = (np.abs(d_lon[0]) / 360) * (area_lat_btm - area_lat_top)
+    area = np.where(area < 0, np.nan, area)    # Prevent negative values at latitude of -90 degrees
```

Comment:
No need to raise a ValueError because there will always be a row of nan values when the latitude is -90°, and we prefer to keep the xarray coordinates to be consistent with the inputs `lon_coords` and `lat_coords`